src/game_battlealgorithm.cpp

```diff
@@ -1135,68 +1148,91 @@ bool Game_BattleAlgorithm::Skill::Execute() {
-			if (skill.affect_hp) {
-				if (Player::IsRPG2k3()) {
-					this->hp = effect;
-				} else {
-					this->hp = std::max<int>(0, std::min<int>(effect, GetTarget()->GetMaxHp() - GetTarget()->GetHp()));
-				}
-			}
-			if (skill.affect_sp && Utils::PercentChance(to_hit)) {
-				int sp_cost = GetSource() == GetTarget() ? source->CalculateSkillCost(skill.ID) : 0;
-				this->sp = std::max<int>(0, std::min<int>(effect, GetTarget()->GetMaxSp() - GetTarget()->GetSp() + sp_cost));
-			}
-			if (skill.affect_attack && Utils::PercentChance(to_hit))
-				this->attack = std::max<int>(0, std::min<int>(effect, std::min<int>(GetTarget()->MaxStatBattleValue(), GetTarget()->GetBaseAtk() * 2) - GetTarget()->GetAtk()));
-			if (skill.affect_defense && Utils::PercentChance(to_hit))
-				this->defense = std::max<int>(0, std::min<int>(effect, std::min<int>(GetTarget()->MaxStatBattleValue(), GetTarget()->GetBaseDef() * 2) - GetTarget()->GetDef()));
-			if (skill.affect_spirit && Utils::PercentChance(to_hit))
-				this->spirit = std::max<int>(0, std::min<int>(effect, std::min<int>(GetTarget()->MaxStatBattleValue(), GetTarget()->GetBaseSpi() * 2) - GetTarget()->GetSpi()));
-			if (skill.affect_agility && Utils::PercentChance(to_hit))
-				this->agility = std::max<int>(0, std::min<int>(effect, std::min<int>(GetTarget()->MaxStatBattleValue(), GetTarget()->GetBaseAgi() * 2) - GetTarget()->GetAgi()));
-
-			this->success = GetAffectedHp() != -1 || GetAffectedSp() != -1 || GetAffectedAttack() > 0
-				|| GetAffectedDefense() > 0 || GetAffectedSpirit() > 0 || GetAffectedAgility() > 0;
 
-			// If resurrected and no HP selected, the effect value is a percentage:
-			if (IsRevived() && !skill.affect_hp) {
-				this->hp = Utils::Clamp(GetTarget()->GetMaxHp() * effect / 100, 1, GetTarget()->GetMaxHp() - GetTarget()->GetHp());
-				this->success = true;
-			}
+		// Handle negative effect from attributes
+		if (effect < 0) {
+			this->negative_effect = true;
+			this->healing = !this->healing;
+			effect = -effect;
 		}
-		if (!healing) {
-			absorb = skill.absorb_damage;
 
-			if (skill.affect_hp) {
+		effect = Utils::Clamp(effect, 0, MaxDamageValue());
+
+		if (IsNegativeSkill()) absorb = skill.absorb_damage;
+
+		if (skill.affect_hp && Utils::PercentChance(to_hit)) {
+			if (IsNegativeSkill()) {
 				this->hp = Algo::AdjustDamageForDefend(effect, *GetTarget());
 
+				// RPG_RT bug: Negative effects affect HP double
+				if (this->negative_effect) {
+					this->hp *= 2;
+				}
+
 				if (IsAbsorb())
 					this->hp = std::min<int>(hp, GetTarget()->GetHp());
 
-				if (GetTarget()->GetHp() - this->hp <= 0) {
+				if (!this->negative_effect && GetTarget()->GetHp() - this->hp <= 0) {
 					// Death state
 					lethal = true;
 					killed_by_dmg = true;
 				}
+			} else {
+				if (!this->negative_effect) {
+					if (Player::IsRPG2k3()) {
+						this->hp = effect;
+					} else {
+						this->hp = Utils::Clamp(GetTarget()->GetMaxHp() - GetTarget()->GetHp(), 0, effect);
```

Comment:
All these Clamps are written wrong, it should be (effect, 0, maxvalue)

Reply:
This is probably a dumb question but what value I'm supposed to use for maxvalue? Is it MaxDamageValue() or the maximum positive integer value?

Reply:
The maxvalue depends on the context. You just need to flip the order of the arguments here. While what you wrote technically works, it doesn't read well.

The function signature is `Clamp(value, min, max)`

So here it would be:

```
Clamp(effect, 0, GetTarget()->GetMaxHp() - GetTarget()->GetHp())
```

Reply:
The order of the arguments have been corrected now.